pandas/core/indexes/base.py

```diff
@@ -4281,22 +4291,26 @@ def _join_non_unique(
         # We only get here if dtypes match
         assert self.dtype == other.dtype
 
-        lvalues = self._get_join_target()
-        rvalues = other._get_join_target()
-
         left_idx, right_idx = get_join_indexers(
-            [lvalues], [rvalues], how=how, sort=True
+            [self._values], [other._values], how=how, sort=True
         )
+        mask = left_idx == -1
 
-        left_idx = ensure_platform_int(left_idx)
-        right_idx = ensure_platform_int(right_idx)
+        # error: Argument 1 to "take" of "ExtensionArray" has incompatible
+        # type "ndarray[Any, dtype[signedinteger[Any]]]"; expected "Sequence[int]"
+        join_array = self._values.take(left_idx)  # type: ignore[arg-type]
+        # error: Argument 1 to "take" of "ExtensionArray" has incompatible type
+        # "ndarray[Any, dtype[signedinteger[Any]]]"; expected "Sequence[int]"
+        right = other._values.take(right_idx)  # type: ignore[arg-type]
 
-        join_array = np.asarray(lvalues.take(left_idx))
-        mask = left_idx == -1
-        np.putmask(join_array, mask, rvalues.take(right_idx))
+        # TODO: just use Index.putmask once GH#43212 is fixed
```

Comment:
is this changable?  as merged the PR around that

Reply:
good catch, will update